File: Top_Interview_Questions/22_GenerateParentheses.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> ans;
        dfs(n, n, "", ans);
        return ans;
    }
private:
    void dfs(int left, int right, string cur, vector<string>& ans) {
        if(left > right) return;
        if(left == 0 && right == 0) {
            ans.push_back(cur);
            return;
        }
        if(left > 0)
            dfs(left - 1, right, cur + '(', ans);
        if(right > 0)
            dfs(left, right - 1, cur + ')', ans);
    }
};

File: Question_Java/22_Generate_Parentheses.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> ans = new ArrayList<>();
        dfs(n, n, new StringBuilder(), ans);
        return ans;
    }
    void dfs(int left, int right, StringBuilder cur, List<String> ans) {
        if (left < 0 || right < 0) return;
        if (left == 0 && right == 0) {
            ans.add(cur.toString());
            return;
        }
        if (right < left) return;

        cur.append('(');
        dfs(left - 1, right, cur, ans);
        cur.deleteCharAt(cur.length() - 1);
        cur.append(')');
        dfs(left, right - 1, cur, ans);
        cur.deleteCharAt(cur.length() - 1);
    }
}